endpoints/oauth/login.py

```
@@ -117,9 +117,15 @@ def callback_func():
 
         # Exchange the OAuth code for login information.
         code = request.values.get("code")
+        kwargs = {}
+        if hasattr(login_service, "pkce_enabled") and login_service.pkce_enabled():
+            session_key = f"_oauth_pkce_{login_service.service_id()}"
+            data = session.pop(session_key, None)
+            if data and hasattr(data, "get") and data.get("verifier"):
```

this check hasattr(data, "get") before accessing data.get("verifier") can potentially be improved for defensive programming and code clarity. 
Since session.pop() should return either a dict or None, it may be clearer and safer to check 
`if isinstance(data, dict) and "verifier" in data:`

This avoids silent failures if data is not as expected.
There are a couple of other places where this same improvement can be made.